Reqnroll.Generator/Generation/UnitTestFeatureGenerator.cs

```diff
@@ -272,29 +284,79 @@ private void SetupTestInitializeMethod(TestClassGenerationContext generationCont
+                    new CodePropertyReferenceExpression(
+                        featureContextExpression,
+                        nameof(FeatureContext.BeforeFeatureHookFailed)),
+                    [new CodeThrowExceptionStatement(
+                        new CodeObjectCreateExpression(
+                            new CodeTypeReference(typeof(ReqnrollException), CodeTypeReferenceOptions.GlobalReference),
+                            new CodePrimitiveExpression("Scenario skipped because of previous before feature hook error")))]);
+
+            // Will generate this for C#:
+            // finally {
+            //   if (testRunner.FeatureContext == null) { // "Start" the feature if needed
+            //     await testRunner.OnFeatureStartAsync(featureInfo);
+            //   }
+            // }
+            CodeStatement onFeatureStartExpression;
+            var featureStartMethodInvocation = new CodeMethodInvokeExpression(
+                    testRunnerField,
+                    nameof(ITestRunner.OnFeatureStartAsync),
+                    new CodeVariableReferenceExpression(GeneratorConstants.FEATUREINFO_FIELD));
+            if (_codeDomHelper.TargetLanguage != CodeDomProviderLanguage.VB)
+            {
+                _codeDomHelper.MarkCodeMethodInvokeExpressionAsAwait(featureStartMethodInvocation);
+                onFeatureStartExpression = new CodeExpressionStatement(featureStartMethodInvocation);
+            }
+            else
+            // will generate this for VB:
+            // Finally
+            //   If testRunner.FeatureContext Is Nothing Then
+            //     onFeatureStartTask = testRunner.OnFeatureStartAsync(featureInfo)
+            //   EndIf
+            // EndTry
+            // If onFeatureStartTask IsNot Nothing Then
+            //   Await onFeatureStartTask
+            // EndIf
+            {
+                onFeatureStartExpression = new CodeAssignStatement(
+                    new CodeVariableReferenceExpression("onFeatureStartTask"),
+                    featureStartMethodInvocation);
+            }
 
-            testInitializeMethod.Statements.Add(
+            var conditionallyExecuteFeatureStartExpressionStatement =
                 new CodeConditionStatement(
                     new CodeBinaryOperatorExpression(
                         featureContextExpression,
                         CodeBinaryOperatorType.IdentityEquality,
                         new CodePrimitiveExpression(null)),
-                    new CodeExpressionStatement(
-                        onFeatureStartExpression)));
+                    onFeatureStartExpression);
+
+            testInitializeMethod.Statements.Add(
+                new CodeTryCatchFinallyStatement(
+                    [conditionallyExecuteOnFeatureEndExpressionStatement],
+                    [],
+                    [conditionallyExecuteFeatureStartExpressionStatement, throwErrorOnPreviousFeatureStartError]));
```

Comment:
How will throwErrorOnPreviouslFeatureStartError interact with VB generated code? This will throw an exception after a Task has been created, but not yet awaited upon (which is after the Finally clause).

Reply:
Moved the throwErrorOnPreviousFeatureStartError up above the invocation of the FeatureStart.
I believe a truth-table for what we want as an outcome is the following. Let me know if you think otherwise:
| FeatureContext Instance| BeforeFeatureHookFailed| OnFeatureStartThrows| Scenario is Executed?|
|--------|--------|--------|--------|
| Null|  not checked | FALSE| Yes|
| Null|  not checked | TRUE| No|
| Not Null| FALSE| FALSE| Yes|
| Not Null| FALSE| TRUE| No|
| Not Null| TRUE| FALSE| No| 
| Not Null| TRUE| TRUE| No|

Reply:
The decision table is even simpler, because the `OnFeatureStart` is only relevant for the first scenario execution of a feature (FeatureContext is null), while the BeforeFeatureHookFailed is an indicator for the subsequent scenario executions of the same feature (FeatureContext is not null). So the table is:

| FeatureContext Instance| BeforeFeatureHookFailed| OnFeatureStartThrows| Scenario is Executed?|
|--------|--------|--------|--------|
| Null (first scenario of feature) |  not checked | FALSE| Yes|
| Null  (first scenario of feature) |  not checked | TRUE| No|
| Not Null (further scenario in feature) | FALSE| OnFeatureStart not executed | Yes|
| Not Null (further scenario in feature) | TRUE| OnFeatureStart not executed | No|